# <center><span style="color:red">Cálculo Numérico - Avaliação - 01</span><br/></center>


## <center><span style="color:green">Exercício  00</span><br/></center>


In [1]:
from fractions import Fraction
import math
from sympy import *
import numpy as np
import matplotlib.pyplot as plt

### <center><span style="color:yellow">Calculando matriz de Hilbert de ordem 4</span><br/></center>

In [2]:
def Hilbert_matrix_string(a, b):
    return [[str(Fraction(1 / (i + j + 1)).limit_denominator()) for j in range(b)] for i in range(a)]


Hilbert_order_4_matrix = Hilbert_matrix_string(4,4)
print("Matriz de Hilbert ordem 4:")
Hilbert_order_4_matrix

Matriz de Hilbert ordem 4:


[['1', '1/2', '1/3', '1/4'],
 ['1/2', '1/3', '1/4', '1/5'],
 ['1/3', '1/4', '1/5', '1/6'],
 ['1/4', '1/5', '1/6', '1/7']]

### <center><span style="color:yellow">Calculando matriz inversa</span><br/></center>

In [3]:
def Hilbert_matrix_array(a, b):
    return np.array([[Fraction(1 / (i + j + 1)).limit_denominator() for j in range(b)] for i in range(a)])

hilbert_matrix_array_order_4 = Hilbert_matrix_array(4,4).astype('float64')
Hilbert_inverse_matrix_array_order_4 = np.linalg.inv(hilbert_matrix_array_order_4)
print("Matriz inversa de Hilbert de ordem 4:")
Hilbert_inverse_matrix_array_order_4

Matriz inversa de Hilbert de ordem 4:


array([[   16.,  -120.,   240.,  -140.],
       [ -120.,  1200., -2700.,  1680.],
       [  240., -2700.,  6480., -4200.],
       [ -140.,  1680., -4200.,  2800.]])

### <center><span style="color:yellow">Descrevendo vetor solução usando pinv (pseudo-inverse) function</span><br/></center>

In [4]:
pinv = np.linalg.pinv(hilbert_matrix_array_order_4)
b = [0 ,1, 10, 100]
solution_vector = pinv.dot(b)
print("O vetor solucao é dado por:")
solution_vector.tolist()

O vetor solucao é dado por:


[-11719.999999997157,
 142199.99999996758,
 -357899.99999992223,
 239679.9999999497]

## <center><span style="color:green">Exercício 01 - Parte a: Calculando xm para altura máxima</span><br/></center>

In [5]:
x = symbols('x')

g, y0, v0, m, c = 9.81, 100, 55, 80, 15
y = y0 + (m / c) * (v0 + ((m * g) / c)) * (1 - exp(- (c / m) * x)) - ((m * g) / c) * x
print('Equacao:')
y


Equacao:


-52.32*x + 672.373333333333 - 572.373333333333*exp(-0.1875*x)

## Altura maxima de xm é dada para y'(x) = 0

In [6]:
y_diff = diff(y)
print("y(x)' = ", y_diff)
print("\nValor para y(x)'= 0:")
solve(y_diff)[0]

y(x)' =  -52.32 + 107.32*exp(-0.1875*x)

Valor para y(x)'= 0:


3.83166036484523

## <center><span style="color:green">Parte B - Calculando x para y(x) = 0</span><br/></center>

In [7]:
print("Valor de y(x) = 0: ")
solve(y, rational = False)[0]

Valor de y(x) = 0: 


11.6108384710610

### <center><span style="color:yellow">O processo deve ser iterativo, pois a equacao -52.32 * x + 672.373333333333 - 572.373333333333 * exp(-0.1875 * x) nao possui solucao elementar.</span><br/></center>

## <center><span style="color:green">Exercicio 02 parte a - Resolvendo o primeiro sistema de equacoes</span><br/></center>

In [8]:
x, y = symbols('x y')
# teste
nsolve([x+y**2-4, exp(x)+x*y-3], [x, y], [1, 1])

Matrix([
[0.620344523485226],
[ 1.83838393066159]])

In [ ]:
x, y = symbols('x y')
# equacao real
nsolve([x**2 + y**2, -exp(x) + y], [x, y], [-1, -1])

## <center><span style="color:green">Exercicio 03 - descobrindo valor de d</span><br/></center>

In [ ]:
d = symbols('d')
k1, k2, m, g, h = 40000, 40, 95, 9.81, 0.43 
f = -(k1 * d + k2 * d ** (3/2))
energy_conservation = (2 * k2 * d ** (5/2))/5 + (1/2) * k1 * d ** 2 - m *g * d - m * g *h
energy_conservation
solve(energy_conservation, rational = False)[0]
